### PR TITLE
Update `fileWriteCsvFromStream` to maintain the order while writing

### PR DESCRIPTION
**Pull request by sahanHe**
## Purpose
This will fix the issue related to maintaining field order in `fileWriteCsvFromStream()` API
Fixes: [3399](https://github.com/ballerina-platform/ballerina-standard-library/issues/3399)
## Checklist
- [x] Linked to an issue
- [x] Updated the changelog
- [x] Added tests
- [x] Updated the spec
- [x] Checked native-image compatibility

### CODE REVIEW

**Comment by daneshk** [comment]
@sahanHe Please link the issue here. And add a description

File: ballerina/channel_csv_io.bal [hunk 1]
```diff
@@ -60,35 +60,64 @@ isolated function channelWriteCsv(string path, FileWriteOption option, string[][
...
-isolated function channelWriteCsvFromStream(WritableChannel writableChannel, stream<string[]|map<anydata>, Error?> csvStream) returns
+isolated function channelWriteCsvFromStream(string path, FileWriteOption option, stream<string[]|map<anydata>, Error?> csvStreamToWrite) returns
 Error? {
-    WritableCSVChannel csvChannel = check getWritableCSVChannel(writableChannel);
-    do {
-        if csvStream is stream<string[], Error?> {
-            record {|string[] value;|}|Error? csvRecordString = csvStream.next();
+    string[] headersFromCSV = [];
+    string[] headersFromStream = [];
+    boolean skipHeaders = false;
```
Comment (daneshk) [inline ballerina/channel_csv_io.bal:67]: do we need to initialize them at the beginning

File: ballerina/channel_csv_io.bal [hunk 2]
```diff
@@ -60,35 +60,64 @@ isolated function channelWriteCsv(string path, FileWriteOption option, string[][
...
+    if csvStreamToWrite is stream<string[], Error?> {
+        WritableCSVChannel csvChannel = check getWritableCSVChannel(check openWritableCsvFile(path, option = option));
+        do {
+            record {|string[] value;|}|Error? csvRecordString = csvStreamToWrite.next();
```
Comment (daneshk) [inline ballerina/channel_csv_io.bal:71]: are we handling the error return here?

File: ballerina/channel_csv_io.bal [hunk 3]
```diff
@@ -60,35 +60,64 @@ isolated function channelWriteCsv(string path, FileWriteOption option, string[][
...
+        } else {
+            headersFromCSV = check readHeadersFromCsvFile(path);
+            do {
+                csvRecordMap = check csvStreamToWrite.next();
+                if csvRecordMap !is () {
+                    headersFromStream = csvRecordMap["value"].keys();
+                } else {
+                    check csvStreamToWrite.close();
+                    return;
+                }
```
Comment (daneshk) [inline ballerina/channel_csv_io.bal:107]: These lines are same for both Override and Append

File: ballerina/channel_csv_io.bal [hunk 4]
```diff
@@ -186,6 +215,31 @@ isolated function writeRecordsToCsvFile(string path, map<anydata>[] contentToWri
...
 
+
+isolated function writeRecordsStreamToCsvFile(string path, stream<map<anydata>, Error?> csvStreamToWrite, string[] headers, FileWriteOption option, boolean skipHeaders, 
+    record {|map<anydata> value;|}? csvRecord) returns Error? {
+    WritableCSVChannel csvChannel = check getWritableCSVChannel(check openWritableCsvFile(path, option = option));
+    do {
+        if !skipHeaders {
+            check csvChannel.write(headers);
+        }
+        record {|map<anydata> value;|}? csvRecordMap = csvRecord;
```
Comment (daneshk) [inline ballerina/channel_csv_io.bal:226]: Do we need this line

Comment (daneshk) [inline ballerina/channel_csv_io.bal:226]: what is csvRecord here?

File: ballerina/channel_csv_io.bal [hunk 5]
```diff
@@ -186,6 +215,31 @@ isolated function writeRecordsToCsvFile(string path, map<anydata>[] contentToWri
...
+        while csvRecordMap is record {|map<anydata> value;|} {
+            string[] sValues = [];
+            foreach string key in headers {
+                sValues.push(csvRecordMap["value"].get(key).toString());
+            }
+            csvRecordMap = check csvStreamToWrite.next();
+            check csvChannel.write(sValues);
+        }
+        } on fail Error err {
```
Comment (daneshk) [inline ballerina/channel_csv_io.bal:235]: formatting issue.

File: ballerina/channel_csv_io.bal (hunk 3 above)
Comment (sahanHe) [reply ballerina/channel_csv_io.bal]: updated in [c475f59](https://github.com/ballerina-platform/module-ballerina-io/pull/830/commits/c475f5979f3f0d6e8f7b7b160c9f54d824a33577)

File: ballerina/channel_csv_io.bal [hunk 6]
```diff
@@ -52,43 +52,67 @@ isolated function channelWriteCsv(string path, FileWriteOption option, string[][
...
+                csvRecordString = csvStreamToWrite.next();
+            }
+        } on fail Error err {
+            check csvStreamToWrite.close();
+            check csvChannel.close();
+            return err;
+        }
+        check csvChannel.close();
+    } else if csvStreamToWrite is stream<map<anydata>, Error?> {
+        WritableCSVChannel csvChannel = check getWritableCSVChannel(check openWritableCsvFile(path, option = option));
```
Comment (daneshk) [inline ballerina/channel_csv_io.bal:80]: Is it correct to create a channel here? what if there is nothing to write in the stream?

**Review by niveathika — COMMENTED** [review]
PR check list need to be followed. As well as spec and API doc changes related to this

File: ballerina/channel_csv_io.bal (hunk 2 above)
Comment (niveathika) [reply ballerina/channel_csv_io.bal:68]: Same question. From the logic seems its ignored. Is it the correct way?

File: ballerina/channel_csv_io.bal [hunk 7]
```diff
@@ -52,43 +52,74 @@ isolated function channelWriteCsv(string path, FileWriteOption option, string[][
...
+        } on fail Error err {
+            check csvStreamToWrite.close();
+            check csvChannel.close();
+            return err;
+        }
+        check csvChannel.close();
+    } else if csvStreamToWrite is stream<map<anydata>, Error?> {
+        boolean skipHeaders = false;
+        string[] headersFromStream;
+        record {|map<anydata> value;|}? csvRecordMap = check csvStreamToWrite.next();
```
Comment (niveathika) [inline ballerina/channel_csv_io.bal:82]: If error is returned from check, wouldn't the channel be closed properly here?

File: ballerina/channel_csv_io.bal [hunk 8]
```diff
@@ -52,43 +52,74 @@ isolated function channelWriteCsv(string path, FileWriteOption option, string[][
...
-            record {|map<anydata> value;|}? csvRecordMap = check csvStream.next();
             while csvRecordMap is record {|map<anydata> value;|} {
                 string[] sValues = [];
-                foreach [string, anydata] [_, value] in csvRecordMap["value"].entries() {
-                    sValues.push(value.toString());
+                foreach string header in headersFromStream {
+                    sValues.push(csvRecordMap["value"].get(header).toString());
                 }
+                csvRecordMap = check csvStreamToWrite.next();
                 check csvChannel.write(sValues);
```
Comment (niveathika) [inline ballerina/channel_csv_io.bal:113]: Is this the correct order? We are iterating the stream even before the first element write.

File: docs/spec/spec.md [hunk 9]
```diff
@@ -306,27 +306,27 @@ The following API writes given CSV content to a given file. When writing a `Reco
...
+# stream<record{}, io:Error?>|io:Error content = io:fileReadCsvAsStream("./resources/myfile.csv");
 # ```
 # + path - The CSV file path
-# + return - The entire CSV content in the channel a stream of string arrays or an `io:Error`
+# + return - The entire CSV content in the channel a stream of string arrays, ballerina records or an `io:Error`
 public isolated function fileReadCsvAsStream(string path) returns stream<string[], Error?>|Error;
 ```
 
-The following API writes a given CSV stream to a given file.
+The following API writes a given CSV stream to a given file. When writing a `Record[]` content to a CSV file in `OVERWRITE`, by default, headers will be written to the CSV file as unlike `string[]` the order of record fields is not guaranteed. For `APPEND`, order of the existing csv file is inferred using the headers and used as the order
```
Comment (niveathika) [inline docs/spec/spec.md:329]:
```suggestion
The following API writes a given CSV stream to a given file. When writing string[] to CSV, the order of the array elements are considered. However when writing a `record{}[]`, the order is inferred from the CSV headers (CSV header should match the fields). Moreover, in OVERWRITE mode fields are written as headers to preserve order between writes.
```

File: changelog.md [hunk 10]
```diff
@@ -3,7 +3,11 @@ This file contains all the notable changes done to the Ballerina I/O package thr
 
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
-## [Unreleased]
+## [Unreleased] 
+### Fixed
```
Comment (niveathika) [inline changelog.md:7]:
```suggestion

### Changed
```

File: docs/spec/spec.md (hunk 9 above)
Comment (niveathika) [reply docs/spec/spec.md:329]: Lets add similar message in API docs. Spec is not the first place a user refers this.

Comment (daneshk) [reply docs/spec/spec.md:330]: @sahanHe did we update the API Docs

File: ballerina/channel_csv_io.bal (hunk 7 above)
Comment (daneshk) [reply ballerina/channel_csv_io.bal:82]: IMO, We need to handle the error and close the stream here.

File: ballerina/channel_csv_io.bal [hunk 11]
```diff
@@ -52,43 +52,70 @@ isolated function channelWriteCsv(string path, FileWriteOption option, string[][
...
+            return;
+        }
+        if option == APPEND {
+            string[] headersFromCSV = check readHeadersFromCsvFile(path);
+            headersFromCSV = check validateCsvHeaders(headersFromCSV, headersFromStream);
+            if headersFromCSV.length() != 0 {
+                skipHeaders = true;
+                headersFromStream = headersFromCSV;
+            }
+        }
```
Comment (daneshk) [inline ballerina/channel_csv_io.bal:98]:
```suggestion
    } else if csvStreamToWrite is stream<map<anydata>, Error?> {
        boolean skipHeaders = false;
        string[] headersFromStream;
        record {|map<anydata> value;|}|Error? csvRecordMap = csvStreamToWrite.next();
        if (csvRecordMap is record {|map<anydata> value;|}) {
            headersFromStream = csvRecordMap["value"].keys();
        } else {
            check csvStreamToWrite.close();
            return;
        }
        if option == APPEND {
            string[] headersFromCSV = check readHeadersFromCsvFile(path);
            headersFromCSV = check validateCsvHeaders(headersFromCSV, headersFromStream);
            if headersFromCSV.length() != 0 {
                skipHeaders = true;
                headersFromStream = headersFromCSV;
            }
        }
        WritableCSVChannel csvChannel = check getWritableCSVChannel(check openWritableCsvFile(path, option = option));
```

Comment (daneshk) [reply ballerina/channel_csv_io.bal:98]: move creating WritableCSVChannel to below, as we are not using

**Review by daneshk — APPROVED** [review]
LGTM

**Pull request merged**